File: 234-palindrome-linked-list/234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if(head == null || head.next == null) return true;
        Stack<ListNode> stack = new Stack<>();
        ListNode f = head, s = head;
        while(s != null && s.next != null) {
            s = s.next.next;
            f = f.next;
        }
        if(s != null) f = f.next;
        
        while(f != null) {
            stack.push(f);
            f = f.next;
        }
        s = head;
        while(!stack.isEmpty()) {
            if(s.val != stack.peek().val) return false;
            s = s.next;
            stack.pop();
        }
        return true;
    }
}

File: 234-palindrome-linked-list/234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        #find mid point of the list, reverse list  and compare
        fast = slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow =slow.next
        if fast: slow = slow.next
        def reverse(head):
            prev, curr = None, head
            while curr:
                next = curr.next
                curr.next = prev
                prev = curr
                curr = next
            return prev
        fast = reverse(slow)
        slow = head
        while fast:
            if fast.val != slow.val:
                return False
            fast = fast.next
            slow =slow.next
        return True
            